File: src/main/java/com/marsraver/LedFx/AnimationType.java
package com.marsraver.LedFx;

import com.marsraver.LedFx.animations.*;

/**
 * Enumeration of available animation types for the LED framework.
 * Each type corresponds to a different visual animation that can be displayed.
 */
public enum AnimationType {
    TEST("test", "Test Animation"),
    SPINNING_BEACHBALL("spinning-beachball", "Spinning Beachball Animation"),
    BOUNCING_BALL("bouncing-ball", "Bouncing Ball Animation"),
    MUSIC_BALL("music-ball", "Music Ball Animation"),
    VIDEO_PLAYER("video-player", "Video Player Animation"),
    FAST_PLASMA("fast-plasma", "Fast Plasma Animation"),
    CLOUDS("clouds", "Clouds Animation"),
    PERLIN_OSCILLATOR("perlin-oscillator", "Perlin Oscillator Animation"),
    STARFIELD("starfield", "Starfield Animation"),
    BLACK_HOLE("black-hole", "Black Hole Animation"),
    BLURZ("blurz", "Blurz Animation");
    
    private final String id;
    private final String displayName;
    
    AnimationType(String id, String displayName) {
        this.id = id;
        this.displayName = displayName;
    }
    
    /**
     * Gets the unique identifier for this animation type.
     * 
     * @return The animation ID
     */
    public String getId() {
        return id;
    }
    
    /**
     * Gets the human-readable display name for this animation type.
     * 
     * @return The display name
     */
    public String getDisplayName() {
        return displayName;
    }
    
    /**
     * Finds an animation type by its ID.
     * 
     * @param id The animation ID to search for
     * @return The matching AnimationType, or null if not found
     */
    public static AnimationType fromId(String id) {
        for (AnimationType type : values()) {
            if (type.id.equals(id)) {
                return type;
            }
        }
        return null;
    }
    
    /**
     * Gets all available animation types as a formatted string.
     * 
     * @return A string listing all available animations
     */
    public static String getAvailableAnimations() {
        StringBuilder sb = new StringBuilder("Available animations:\n");
        for (AnimationType type : values()) {
            sb.append("  ").append(type.id).append(" - ").append(type.displayName).append("\n");
        }
        return sb.toString();
    }


    public static LedAnimation createAnimation(String animationId) {
        AnimationType type = AnimationType.fromId(animationId);
        assert type != null;
        return createAnimation(type);
    }

    public static LedAnimation createAnimation(AnimationType animationType) {
            return switch (animationType) {
                case TEST -> new TestAnimation();
                case SPINNING_BEACHBALL -> new SpinningBeachballAnimation();
                case BOUNCING_BALL -> new BouncingBallAnimation();
                case MUSIC_BALL -> new MusicBallAnimation();
                case VIDEO_PLAYER -> new VideoPlayerAnimation();
                case FAST_PLASMA -> new FastPlasmaAnimation();
                case CLOUDS -> new CloudsAnimation();
                case PERLIN_OSCILLATOR -> new PerlinOscillatorAnimation();
                case STARFIELD -> new StarfieldAnimation();
                case BLACK_HOLE -> new BlackHoleAnimation();
                case BLURZ -> new BlurzAnimation();
            };
    }
}

File: src/main/kotlin/com/marsraver/LedFx/AnimationType.kt
package com.marsraver.LedFx

import com.marsraver.LedFx.animations.*

/**
 * Enumeration of available animation types for the LED framework.
 * Each type corresponds to a different visual animation that can be displayed.
 */
enum class AnimationType(
    /**
     * Gets the unique identifier for this animation type.
     *
     * @return The animation ID
     */
    val id: String,
    displayName: String
) {
    AKEMI("akemi", "Akemi Animation"),
    AUDIO_VISUALIZER("audio-visualizer", "Audio Visualizer"),
    BLACK_HOLE("black-hole", "Black Hole Animation"),
    BLURZ("blurz", "Blurz Animation"),
    BOUNCING_BALL("bouncing-ball", "Bouncing Ball Animation"),
    CLOUDS("clouds", "Clouds Animation"),
    FALLING_SNOW("falling-snow", "Falling Snow Animation"),
    FAST_PLASMA("fast-plasma", "Fast Plasma Animation"),
    MUSIC_BALL("music-ball", "Music Ball Animation"),
    PERLIN_OSCILLATOR("perlin-oscillator", "Perlin Oscillator Animation"),
    SOUND_BUBBLE("sound-bubble", "Sound Bubble Animation"),
    SPINNING_BEACHBALL("spinning-beachball", "Spinning Beachball Animation"),
    STARFIELD("starfield", "Starfield Animation"),
    TRON_RECOGNIZER("tron-recognizer", "Tron Recognizer Animation"),
    VIDEO_PLAYER("video-player", "Video Player Animation");

    /**
     * Gets the human-readable display name for this animation type.
     */
    val displayName: String = displayName

    override fun toString(): String = displayName

    companion object {
        /**
         * Finds an animation type by its ID.
         *
         * @param id The animation ID to search for
         * @return The matching AnimationType, or null if not found
         */
        fun fromId(id: String?): AnimationType? {
            for (type in AnimationType.entries) {
                if (type.id == id) {
                    return type
                }
            }
            return null
        }

        val availableAnimations: String
            /**
             * Gets all available animation types as a formatted string.
             *
             * @return A string listing all available animations
             */
            get() {
                val sb = StringBuilder("Available animations:\n")
                for (type in AnimationType.entries) {
                    sb.append("  ").append(type.id).append(" - ").append(type.displayName).append("\n")
                }
                return sb.toString()
            }

        fun createAnimation(animationType: AnimationType): LedAnimation =
            when (animationType) {
                AKEMI -> AkemiAnimation()
                AUDIO_VISUALIZER -> AudioVisualizerAnimation()
                BLACK_HOLE -> BlackHoleAnimation()
                BLURZ -> BlurzAnimation()
                BOUNCING_BALL -> BouncingBallAnimation()
                CLOUDS -> CloudsAnimation()
                FALLING_SNOW -> FallingSnowAnimation()
                FAST_PLASMA -> FastPlasmaAnimation()
                MUSIC_BALL -> MusicBallAnimation()
                PERLIN_OSCILLATOR -> PerlinOscillatorAnimation()
                SOUND_BUBBLE -> SoundBubbleAnimation()
                SPINNING_BEACHBALL -> SpinningBeachballAnimation()
                STARFIELD -> StarfieldAnimation()
                TRON_RECOGNIZER -> TronRecognizerAnimation()
                VIDEO_PLAYER -> VideoPlayerAnimation()
            }
    }
}
